### DISCUSSION TITLE
my 1st test of allskyannotate

### DISCUSSION THREAD

**Discussion by AndreasLMeg** [Alex-developer/allskyannotate#1] [General]
@Alex-developer 
Hello Alex,

I am very interested in your implementation and will try it once.
I will report all anomalies here. But I don't want to spam you with issues.
1. Thank you for your work, looks professional and good
2. Windows/UNIX Format/CRLF line terminators - fixed with dos2unix
```
pi@allsky:~/allskyannotate $ ./installer/install.sh
-bash: ./installer/install.sh: /bin/bash^M: Defekter Interpreter: Datei oder Verzeichnis nicht gefunden
```
3. install.sh: got this message - (in the first step I will not change anything)
```
  The scripts f2py, f2py3 and f2py3.7 are installed in '/home/pi/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
```
4. install.sh: after "Installation complete." the window " AllSky Image Annotater Installer" is still visible
5. saveImageNight.sh: path wrong? - changed to "/home/pi/allskyannotate/annotate.py"
```
# Run the image annotation script. Added by the annotater installer on Di Dez 28 06:41:02 CET 2021
if [[ -x "/home/pi/allskyannotate/annotate.py" ]]
then
  /home/pi/allskyannotate/annotate.py
fi
``` 
6. CR/LF
```
pi@allsky:~/allskyannotate $ file annotate.py
annotate.py: Python script, ASCII text executable, with CRLF line terminators

Dec 28 06:55:58 allsky allsky.sh[16788]: scripts/saveImageNight.sh: /home/pi/allskyannotate/annotate.py: /usr/bin/python3^M: Defekter Interpreter: Datei oder Verzeichnis nicht gefunden

pi@allsky:~/allskyannotate $ dos2unix annotate.py
dos2unix: Datei annotate.py wird ins Unix-Format umgewandelt …
```

Result: (my allsyky software includes some tests with extratext and umlauts)
![grafik](https://user-images.githubusercontent.com/85451655/147537949-70984159-24d6-4e69-8e4e-e332ce850961.png)

Conclusion:
1. Only a few adjustments necessary
2. looks very well thought out and useful
3. very good if Allsky can concentrate on the main tasks
4. Very robust - also configurable for me
5. Incorrect configurations are nicely displayed in the logging
6. Very good job

**Comment by Alex-developer** [comment]
@AndreasLMeg 

Thanks for the feedback really appreciate it :-)

Some feedback on the points

1) No problem. I have been meaning to do this for ages
2) That appears as if you don't have bash available?
3) I have fixed this and pushed a change to master
4) I have fixed this and pushed a change to master, it was waiting for a confirmation on an apt install !!!!
5) Are you saying the install path in the script as wrong? This is not something I have seen. Did you run the install script from inside the installer folder?
6) I dont see this, see below for the tests I have don on buster

I also think it would be a good idea to had a hook script in the same script so that I don't have to modify it, this will of course help when its changed to a single saveimage script. Happy to do a PR for this if you want?

To test all of the above having fixed a couple of bugs in the installer I have

1) Installed a fresh copy of buster (Linux raspberrypi 5.10.63-v7l+ #1496 SMP Wed Dec 1 15:58:56 GMT 2021 armv7l GNU/Linux)
2) Enabled ssh as I work remotely
3) Enable the Pi Camera
4) Installed allsky using a PI Camera
5) Install allsky gui
6) Verify images being captured
7) Install annotate
8) Configured annotate
9) Verified its annotating the images ok
10) SWitched to a ZWO camera
11) Verified its working 

Everything worked fine, I have not seen the bash error nor the Windows CRLF issue

file annotate.py
annotate.py: Python script, ASCII text executable


Now going to repeat all of the above on bullseye !

Cheers

Alex

**Reply by AndreasLMeg** [reply]
> I also think it would be a good idea to had a hook script in the same script so that I don't have to modify it, this will of course help when its changed to a single saveimage script. Happy to do a PR for this if you want?

yes, would be nice. 

> Everything worked fine, I have not seen the bash error nor the Windows CRLF issue

I will test this - hopefully this year ;-)

thanks
Andreas

**Reply by Alex-developer** [reply]
> > I also think it would be a good idea to had a hook script in the same script so that I don't have to modify it, this will of course help when its changed to a single saveimage script. Happy to do a PR for this if you want?
> 
> yes, would be nice.
> 
> > Everything worked fine, I have not seen the bash error nor the Windows CRLF issue
> 
> I will test this - hopefully this year ;-)
> 
> thanks Andreas

@AndreasLMeg 

Cool, ill look a the the hook script. Plenty of time left this year ;-) I have a rare few days off so plenty of time to code :-)

Oh and do you know if there is a version number available anywhere in the allsky software that I can get to in a script? I can't seem to find one

Alex

**Reply by AndreasLMeg** [reply]
Versioning is the most important thing, but it is usually not done properly.

I only know the output "printf (" *** Allsky Camera Software v0.8.2c | 2021 *** \ n ");"

It would be nice to have a definition that everyone involved can use (exe, scripts, external scripts, web ui, ...)
Unfortunately, the software is not tagged either
This is still a construction site that still bothers me, but I can't help ...

Andreas

**Reply by Alex-developer** [reply]
> Versioning is the most important thing, but it is usually not done properly.
> 
> I only know the output "printf (" *** Allsky Camera Software v0.8.2c | 2021 *** \ n ");"
> 
> It would be nice to have a definition that everyone involved can use (exe, scripts, external scripts, web ui, ...) Unfortunately, the software is not tagged either This is still a construction site that still bothers me, but I can't help ...
> 
> Andreas

@AndreasLMeg 

Andreas

Thanks for that. 

I am going to do it a slightly different way so that it will work regardless of the version of allsky being used.

Do you have any suggestions for the name of the hook script - I am thinking 'saveImage_additionalSteps.sh' with a .repo version in the repo

Cheers

Alex

**Reply by AndreasLMeg** [reply]
@thomasjacquin, @linuxkidd, @EricClaeys: What do you think of this proposal - I don't want to decide on my own (I'm not the mastermind)

**Comment by EricClaeys** [comment]
@Alex-developer @AndreasLMeg, the `endOfNight_additionalSteps.repo` file is in allsky/scripts, and contains a couple sentences about renaming it to .sh.  I suggest following that convention, so naming your file `saveImage_additionalSteps.repo` and putting in allsky/scripts.

When will the hook script be called from saveImage.sh?  Before any processing of the image has been done, or after?  I can see a need to call it both time, passing an argument specifying if it's before or after processing.  The script could determine if it wants to ignore the "before" or "after" invocation.

FYI, the capture*.cpp files in the "saveImage.sh" PR pass several arguments to saveImage.sh, including a "DAY/NIGHT" flag, the full pathname of the image (which is now in the format "image-yyyymmddhhmmss.jpg"), the exposure length, temperature, gain, and others.

Versioning is hard - the version of what?  The ZWO .cpp file or the RPiHQ .cpp file or the scripts or the README.md file or something else?  Regardless of what the version number is from, what happens when we update a single file?  Does the version number change?

**Reply by Alex-developer** [reply]
> @Alex-developer @AndreasLMeg, the `endOfNight_additionalSteps.repo` file is in allsky/scripts, and contains a couple sentences about renaming it to .sh. I suggest following that convention, so naming your file `saveImage_additionalSteps.repo` and putting in allsky/scripts.
> 
> When will the hook script be called from saveImage.sh? Before any processing of the image has been done, or after? I can see a need to call it both time, passing an argument specifying if it's before or after processing. The script could determine if it wants to ignore the "before" or "after" invocation.
> 
> FYI, the capture*.cpp files in the "saveImage.sh" PR pass several arguments to saveImage.sh, including a "DAY/NIGHT" flag, the full pathname of the image (which is now in the format "image-yyyymmddhhmmss.jpg"), the exposure length, temperature, gain, and others.
> 
> Versioning is hard - the version of what? The ZWO .cpp file or the RPiHQ .cpp file or the scripts or the README.md file or something else? Regardless of what the version number is from, what happens when we update a single file? Does the version number change?

@EricClaeys 

Thanks for the reply.

I was planning to call the script just before the final image is copied into the daily directory. I agree it makes sense to call it twice. I have the changes ready to make a PR so will add a call to the new script before creating the PR.

I'll take a look at he options passed out from the capture script as it would be handy to have those available.

The version number was more for a simple way for me to check if the annotater installer needed to modify the saveimageNight script or the new hook script. In the end I did it a different way and if the endOfNight_additionalSteps.repo or endOfNight_additionalSteps.sh scripts exist it modifies the hook script otherwise modifies the saveimageNight script.

For all of my professional work I use releases, a planned bunch of changes will lead to a new release being created. The release contains a version number, typically in a text file, that can be read. I agree its impossible to do versioning on a file by file basis.

Cheers

Alex

**Reply by Alex-developer** [reply]
@EricClaeys Quick question

Do you have any preference as to whether the variables should be created in the variables.sh script and then let source "${ALLSKY_HOME}/variables.sh" make them available. This approach is far cleaner than using a load of $1 $2 etc

For now I have just done the ZWO variables as I see the Pi Camera variables are different.

If you are happy with it then I have done all of the code locally and will create a PR, once the saveimage.sh script is done I will do a PR for that as it can use the same hook script

Thanks

Alex

**Comment by EricClaeys** [comment]
@Alex-developer (CC'd @AndreasLMeg)
Which variables are you referring to in your first two paragraphs?  Ones the capture programs pass to saveImage.sh like TEMPERATURE?  variables.sh contains directory-related variables like ALLSKY_HOME, and some other variables most scripts use.  Users normally won't every edit it.  If there are any variables the user needs to set in order to use your script, they should go in the "Overlay" section of the settings_*.json files.

FYI, we plan to merge all the configuration files (settings_*.json, config.sh, ftp_settings.sh) into one large .json file that supports multiple cameras.  Having a single file will make your script a little easier.

The PR for saveImage.sh is being reviewed.  The two capture programs pass arguments to saveImage.sh which will then export them.  You can view the [saveImage.sh branch here](https://github.com/thomasjacquin/allsky/tree/saveImages.sh).  Because there were quite a few changes needed to implement passing arguments from the capture program, I wanted to wait for saveImage.sh so there would only be one script to edit.

The PR will pass these items to saveImage.sh:
* "DAY" or "NIGHT" as $1, exported as $DAY_OR_NIGHT
* The full pathname to the file it just wrote as $2, exported as $CURRENT_IMAGE.  The image name includes the DATE/TIME.
* **[the following variables are passed on the saveImage.sh command line as "variable=value" and exported by saveImage.sh with "THIS_" prefixed to their name, e.g., $THIS_EXPOSURE_US]**
* EXPOSURE_US
* BRIGHTNESS
* MEAN (histogram mean)
* AUTOEXPOSURE (1 == auto exposure on, 0 == off)
* AUTOGAIN (1 or 0)
* AUTOWB (auto white balance, 1 or 0)
* WBR (red balance)
* WBB (blue balance)
* TEMPERATURE
* GAIN
* BIN (1, 2, 3, or 4)
* FLIP (0==none, 1==horizontal, 2==vertical, 3==both)
* BIT_DEPTH (8, 16, or 24)
I suspect there will be many more variables we'll need.  It's fairly easy to add new variables, but requires a PR.

Once the saveImage.sh PR is merged, I'll do another PR to have saveImage.sh call `saveImage_additionalSteps.sh` if it exists.  It will be similar to endOfNight_additionalSteps.repo and .sh.  It will call saveImage_additionalSteps.sh twice - before processing and after, passing an argument specifying if it's BEFORE or AFTER.

I looked at your script - very nice and thanks much for doing this.  I'd like to request some changes to make it possible to remove the text overlay code in the capture programs, and make it easier for users to specify what goes where, and to add some bells and whistles we've been wanting to add.
See the attached file.  The top part is the list of requested changes, followed by a mini-design.  Comments are welcomed.

[Overlay-Alex.txt](https://github.com/Alex-developer/allskyannotate/files/7795367/Overlay-Alex.txt)

**Reply by Alex-developer** [reply]
@EricClaeys 

I was referring to how the variables are passed into the additional steps script but having looked at the code in the PR its not an issue as I see how you are doing it now.

I have been through the document and added some comments. Agree with all of it but just need a bit of clarification on the 'field values' point, sorry probably me just being thick !

Would it make sense to add date and time to the list of exported variables? It would be easier that ripping the filename apart

What I will do is grab the saveimage branch and start moving all of the code into that, adding in the changes. I appreciate this branch may change a bit but its going to be easier to do it

FYI I have a local branch of allsky with the additional file in it for the existing day and night save scripts. Mine looked like this

```
#!/bin/bash

# This script is a hook to run additional operations after an image is captured.
# This script is run before and after the image processing occurs. A variable is
# available, IMAGE_PROCESSING_STAGE, to indicate which stage the script is 
# being called from. You should add your code inside the relevant if statements
# below

# To use this script, first copy it to
# $ALLSKY_HOME/scripts/saveImage_additionalSteps.sh, and then edit the copy.
# This will prevent your changes from being lost in future upgrades.

# Include "${ME}" in any output so it's easier to find in the log file, for debugging.
ME="$(basename "$BASH_ARGV0")"

# Run any required code before image processing starts
if [[ ${IMAGE_PROCESSING_STAGE} == "PRE" ]]; then
fi

# Run any code after image processing has been completed
if [[ ${IMAGE_PROCESSING_STAGE} == "POST" ]]; then
fi
```
I am calling this at two points in the save night and day scripts

```
# Make sure the image isn't corrupted
identify "$IMAGE_TO_USE" >/dev/null 2>&1
RET=$?
if [ $RET -ne 0 ] ; then
	echo -e "${RED}*** ${ME}: ERROR: Image '${IMAGE_TO_USE}' is corrupt; ignoring.${NC}"
	exit 3
fi

IMAGE_PROCESSING_STAGE="PRE"
# Run custom script after the main image processing
cmd="${ALLSKY_SCRIPTS}/saveImage_additionalSteps.sh"
test -x "${cmd}" && "${cmd}"

# Resize the image if required
if [[ $IMG_RESIZE == "true" ]]; then
```

and

```
# Stretch the image
if [[ $AUTO_STRETCH == "true" ]]; then
	[ "${ALLSKY_DEBUG_LEVEL}" -ge 4 ] && echo "${ME}: Stretching '${IMAGE_TO_USE}' by ${AUTO_STRETCH_AMOUNT}"
  	convert "$IMAGE_TO_USE" -sigmoidal-contrast "$AUTO_STRETCH_AMOUNT","$AUTO_STRETCH_MID_POINT" "$IMAGE_TO_USE"
	RET=$?
	if [ $RET -ne 0 ] ; then
		echo -e "${RED}*** ${ME}: ERROR: AUTO_STRETCH failed with RET=${RET}${NC}"
		exit 4
	fi
fi

IMAGE_PROCESSING_STAGE="POST"
# Run custom script after the main image processing
cmd="${ALLSKY_SCRIPTS}/saveImage_additionalSteps.sh"
test -x "${cmd}" && "${cmd}"

# IMG_DIR and IMG_PREFIX are in config.sh
# If the user specified an IMG_PREFIX, copy the file to that name so the websites can display it.
if [ "${IMG_PREFIX}" != "" ]; then
	cp "$IMAGE_TO_USE" "${IMG_PREFIX}${FILENAME}.${EXTENSION}"
fi
```

[Overlay-Alex.txt](https://github.com/Alex-developer/allskyannotate/files/7796732/Overlay-Alex.txt)

**Comment by EricClaeys** [comment]
@Alex-developer 
Cool.  Glad we're on the same page, and thanks for taking the suggestions.

Re: "Field values": In the .json file there are things in quotes before a ":" which in my text file I called "fields" because I don't know the correct term in .json-speak.  Each field has a value which in the .json file is after the ":".  For example:
`"x": 25`
The "field" **x** has a "value" of 25, hence my term "field value".  The "Field Value Types" section of my text file defines what format those values can be.  Examples of a "number" are 25, -25, 25.0, but not +25.
A "percent" is a "number" followed by "%".  The "%" is required.
A "size" can be a number with or without "px", or a percent, or if it has a leading "+" or "-" it's a number relative to the current cursor position.
You nailed "color".
If json limits a "boolean" to "true" and "false" we should use only those; if it doesn't limit a boolean, we should allow yes/no, true/false, and 1/0.

The next section in my text file lists all the fields I could think of that we'd want to support, but I confusingly called them "arguments" because they contain both field names and field value types, e.g.,
`x=xize`
meaning the "field" **x** takes a "field value type" of "size", where "size" is defined in the "Field Value Types" section.

Re: "add date and time to the list of exported variables": probably.  I thought about this prior to my last post, but wasn't sure if people would need the date and time, and if so, would they need the date, or the time, or both, and if so, in what format (YYYYMMDDhhmmss) or the standard UNIX date of number of seconds since 1970 or something else.  So I punted and just said "it's in the filename".

Re: fields with no "variable":  Line 360 in the python script is `if "variable" in fieldData.keys():` which will only execute the lines below it if the "variable" field exists, right?  There is no "else" to that, which I believe mean the "label" won't be displayed if there is no "variable" field.  Is that correct?

Re: fonts.  I'm totally fine with just using True Type fonts and not CV fonts.  I assume we're going to distribute at least one font file with Allsky?  If so, I suggest we put it in allsky/fonts and create $ALLSKY_FONTS in variables.sh  and have the python script look in their if the name of a font does not start with "/", otherwise look where the full pathname points.  Don't look in the scripts directory.

Re: label/variable system:  My vote would be to dump "variable" and call "label" "text" which matches what we call it in the WebUI.

Re: EXIF data: don't do anything with it yet.  It was just a thought.

Re: drag and drop editor:  Cool idea.  Would make a nice addition to the WebUI.

Can you mention the ${CURRENT_IMAGE} and ${DAY_OR_NIGHT} variables in saveImage_additionalSteps.sh?  The first one is the full path name to the image to be modified.  The second one is either "DAY" or "NIGHT" in case someone want to only process an image in daytime or nighttime.  That script should probably also include:
~~~
source "${ALLSKY_HOME}/variables.sh"
source "${ALLSKY_CONFIG}/config.sh"
~~~
so the python script can access $ALLSKY_FONTS, etc.

A nit  In the day and night scripts the first "after" should be "before", just below "`MAGE_PROCESSING_STAGE="PRE"`

**Reply by Alex-developer** [reply]
@EricClaeys Thanks for the fields explanation, makes perfect sense now. Before I start recoding stuff I am going to put together some sample config files just so we get the format agreed.

Regarding the time are you planning for the capture routines to add it to the image or let the overlay script do it? If its the overlay script then unix time is the easiest to work with as its far easier to convert to local formats.

The label code does work its just wrong in my notes. Leaving the variable field in the json but empty will just display the label. It drops into the else on 372 and sets the field value to the variable name, which is empty.

I currently have a default ttf in the repo, its using the same licence as allsky.

Happy to change the name as per the webui.

One issue  I did find last night was that the python script can't see any of the variables in the bash script, just using source will not make them available to it. I did some digging last night and there *may* be a way to do this but if that doesn't work the only way to get them will be to export them from the bash script.

I am back to work on Tuesday so I suspect this will all take a couple of weeks to sort as I will only have evenings

Cheers and happy new year

Alex

**Comment by EricClaeys** [comment]
@Alex-developer Happy New Year to you too.
Good idea with the sample config files.

Once your overlay code is integrated into Allsky I intend to remove all the overlay-related code from the capture*.cpp files so ALL overlay stuff would be done in your script.
I will pass the UNIX time in a "DATETIME" variable.

Does the default.ttf suport non-ASCII characters like the "degree" character?  Ther is an open Issue to support non-ASCII characters. 

Which variables does the python script not see - the ones fro variables.sh and config.sh, or the "THIS_*" variables from saveImage.sh?
The saveImage.sh script exports all the variables from the capture programs so saveImage_additionalSteps.sh should have them.  Is the problem that they aren't getting exported to the python script?  Would doing a "set -a" in saveImage_additionalSteps.sh help?

**Reply by Alex-developer** [reply]
@EricClaeys 

Perfect on the time stuff.

A ttf will support whatever is built into it. Support for non ascii characters will depend on the font author. Best thing we can do is just checks the fonts to see what characters are available. Things like the degree symbol is likely to be supported in most ttf fonts.

The python script will only see variables that are exported from bash so things like ${ALLSKY_CONFIG} are not available. A set -a at the start of the sameImage.sh would be best. I'll test this tomorrow as its a bit late here in the UK now.

I have the saveImage brach locally now so will do some proper testing tomorrow

Cheers

Alex

**Comment by Alex-developer** [comment]
First very basic example; key parts in this example

- Settings section in the json to provide defaults
- Implements a cursor type approach
- Unless overridden settings are inherited from previous field

```
{
    "fields": {
        "temp": {
            "label": "Temperature: ${temp}"
        },
        "exposure": {
            "label": "Exposure: ${exposure}",
            "fonttcolour": "blue"
            
        },
        "sunangle": {
            "label": "Sun Angle ${sunangle}",
            "format": "%.2f"
        },
        "host": {
            "label": "Hostname ${hostname}",
            "position" : {
                "x": 1600,
                "y": 10
            },
            "fontcolour": "white",
            "fontsize": 32
        },
        "gain": {
            "label": "Gain ${gain}"
        }
    },
    "images": {
        "image1": {
            "position": {
                "x": 1820,
                "y": 150
            },
            "rotate": 10,
            "opacity": 0.4,
            "image": "images/compass150.png"            
        }
    },
    "settings": {
        "datafile": "",
        "usecursor": {
            "enabled": true,
            "startx": 10,
            "starty": 10
        },
        "defaultimagetopacity": 0.5,
        "defaultfontopacity": 1,        
        "defaulfonttcolour": "white",
        "defaultfont": {
            "font": "default",
            "size": 48,
            "additionalleading": 0
        }
    },
    "fonts": {
        "moon": {
            "fontPath": "fonts/moon_phases.ttf",
            "fonttcolour": "red",
            "backgroundcolor": "white",
            "fontopacity": 0.5            
        },
        "default": {
            "fontPath": "fonts/Roboto-Medium.ttf"        
        }
    }
}
```
![example1](https://user-images.githubusercontent.com/3010165/148841697-e62b35a7-4136-4ac0-98ac-b21e44207904.png)

**Comment by Alex-developer** [comment]
@EricClaeys Did you see the message I sent on facebook?
